docs/proposals/federated-ruler.md

```diff
@@ -0,0 +1,82 @@
+
+#### Proposal
+
+Since the current default is that a tenant should only be able to write rules against itself, we suggest a config option `ruler.allowed-federated-tenants`, a string slice of OrgIDs like `infra` or `0|1|2|3|4` which are allowed to write rules against all tenants. If a tenant `bar` attempts to create a federated rule, an error should be returned by the ruler api. Similarly an option `ruler.disallowed-federated-tenants` explicitly states a list of tenants for which federated rules are not allowed. Combining these in a `util.AllowedTenants` should allow us to quickly determine if federation is enabled or disabled for a given tenant at rule creation.
+
+### Where to store resulting series of federated rule
+
+#### Challenge
+
+A single tenant rule always stores produced series in the tenant where the rule exists. This 1 -> 1 mapping becomes a many -> 1 mapping for federated rules.
+
+#### Proposal
+
+As some use cases will demand that a specific federated rule, querying tenants `B` and `C`, is stored in the owning team's tenant `A`, an option to allow explicit assignment of destination tenant for a federated rule is needed. Some use cases where a set of tenants `A|B|C|...|Z` are being treated as a single logical tenant when querying this explicit assignment of destination tenant is not needed.
+
+To allow both of these, we suggest an additional field to the rules proto `dest_tenant_id` which determines which tenant to send the series produced by the rule to, used if and only if the rule is a federated rule. If this field is not given, for composite tenants a random but consistent subtenant of the multiple tenants owning the rule is chosen using a hashmod of the series label to determine the subtenant and for single tenants owning a federated rule the resulting series is saved in the tenant which owns the rule.
+
+### Which tenants to query from for federated rules
+
+#### Challenge
+
+A single tenant rule always queries the tenant which owns the rule. This 1 -> 1 mapping becomes a 1 -> many mapping for federated rules.
+
+#### Proposal
+
+As some use cases will demand that a specific federated rule, querying tenant B and C, is stored in the owning teams tenant A, an option to allow explicit assignment of source tenants for a federated rule is needed. In the case of a composite tenant where a set of tenants `A|B|C|...|Z` are being treated as a single logical tenant when querying this explicit assignment of destination tenant isn't explicitly called for, but could prove useful.
+
+To support both of these use cases, we suggest an additional field `source_tenant_ids` on the rule containing an OrgID string e.g. `t0|t1|...|ti` which when present determines which tenants to query for the given rule. In the case of a composite tenant this field would be optional, as ownership of a rule by a composite tenant implies source tenants e.g. for the composite tenant `t0|t1|...|ti` the natural source sub tenants would be `t0`, `t1`, `t2` etc.
```

Comment:
I cheated and have a version of this working that I am cleaning up before opening the PR

The rules looks something like
```
name: record_rule_eg
interval: 10s
rules:
  - record: record_rule_eg
    expr: sum(avalanche_metric_m_4_10017)
    src_tenants: 0|1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19
    dest_tenant: foo
```

Reply:
Ok. Having it in the config makes sense to me. I'm still a bit dubious about `dest_tenant` tho: I'm wondering why it couldn't just be the owner of the config.

Reply:
I'd like to leave `dest_tenant` in to allow an explicit override since we are already extending the prometheus rules structure and it would allow greater flexibility and scalability for federated rules owned by a single tenant.
I'm curious as to @edma2 thoughts on this one

Reply:
IMO it's less of a must-have than controlling the source tenant but since it's optional and defaults to the rule owner tenant I can't think of a reason to not support it unless it introduces weird edge cases or implementation complexity (neither of which I can confidently speak to).

Reply:
At least for the initial implementation I am testing it has not introduced additional complexity not already there from having `src_tenants`

Reply:
Will it be a reasonable requirement that writing evaluated data to multi tenants? Data in each tenant may have different RP. Then `dest_tenant` will be `dest_tenants`.

Reply:
I think that's reasonable @xiaobeiyang 
The logic would remain the same (pick a consistent subtenant out of dest_tenants if dest_tenants is multiple tenants) as the composite tenant case so it'll just be a rename on the implementation PR

Reply:
Opting to remove dest_tenant as there wasn't an explicit use case